### PR TITLE
OEM-> User transfer of devices ownership wizard, triggered by the presence of an empty /boot/oem file

### PR DESCRIPTION
**Pull request by tlaurion**
UPDATE: see [QubesOS OEM to user transfer of ownership](https://github.com/osresearch/heads/pull/507#issuecomment-455814228) below for current state.

Advanced menu addition:
GPG card: factory reset, set key-attr to 4096bits, set passwords, generate keys, remove old keys from rom and insert public and trustdb back, and flash it.
Cryptsetup-reencrypt: reencrypt device and change recovery disk passphrase.

Please review. This is first draft. Everything works but it could be more beautiful and menus should probably be seperated.

@kyle: I was thinking of a oem file being present under /boot to engage a wizard or something.

### CODE REVIEW

File: initrd/etc/functions [hunk 1]
```diff
@@ -217,3 +217,158 @@ preserve_rom() {
...
+    fi
+
+    echo "We will now reencrypt the LUKS container with your own chosen LUKS recovery decryption passphrase."
+    echo "You will be asked to enter the actual recovery LUKS passphrase before choosing your own."
+    echo "CHOOSE IT CAREFULLY AND DO NOT FORGET IT. Good example: white elephant of Antartica"
+    echo "Reencrypting a fresh installation of QubesOS on a 240GB SSD drive can take up to 30 minutes. Be patient."
+    echo -e "\n"
+    cryptsetup-reencrypt -B 32 --use-directio $luks_container
+  else
+    echo "You have to reown the TPM, sign configurations with your GPG smartcard first and select a new default boot option!"
```
Comment (tlaurion) [inline initrd/etc/functions]: @kylerankin : Not ideal. Suggestions?

**Comment by tlaurion** [comment]
@kylerankin @flammit : please comment and suggest improvements in the functional code.

TODO: 
For the GPG part, I was thinking about asking input all at once to the user and store the results of input under /tmp/secret/ files:
- gpg_card_user_password
- gpg_card_admin_password
- gpg_card_full_name
- gpg_card_email
- gpg_card_comment
And then create different menus for:
- factory reset and set-attr
- generate keys and import public and otrust

For the drive reenecryption code, I was thinking into splitting actual code into two submenus:
- LUKS container reencryption (uses current Recovery passphrase and reencrypts everything)
- LUKS Recovery key passphrase change
And then create a meta menu that can do both manually, while oem code calls the functions directly.

File: boards/librem13v2/librem13v2.config [hunk 2]
```diff
@@ -7,6 +7,7 @@ CONFIG_CRYPTSETUP=y
 CONFIG_FLASHROM=y
 CONFIG_FLASHTOOLS=y
 CONFIG_GPG=y
```
Comment (itay-grudev) [inline boards/librem13v2/librem13v2.config]: This line should go.

File: boards/librem15v3/librem15v3.config [hunk 3]
```diff
@@ -9,6 +9,7 @@ CONFIG_CRYPTSETUP=y
 CONFIG_FLASHROM=y
 CONFIG_FLASHTOOLS=y
 CONFIG_GPG=y
```
Comment (itay-grudev) [inline boards/librem15v3/librem15v3.config]: And so do this line.

File: boards/librem13v2/librem13v2.config (hunk 2 above)
Comment (tlaurion) [reply boards/librem13v2/librem13v2.config]: done

File: boards/librem15v3/librem15v3.config (hunk 3 above)
Comment (tlaurion) [reply boards/librem15v3/librem15v3.config]: done

**Comment by tlaurion** [comment]
So at this point, here is what we have. I know, I need a better camera. That will happen soon. :)

**OEM/tech-savvy user common parts:**
- Generate Heads ROM ( in the case of x230: make BOARD=x230-flash && make BOARD=x230). And flash it externally to hardware.
  - x230: Backup 4mb SPI flash chip, reflash x230-flash Heads rom on it, Backup 8mb  SPI flash chip, [apply me_cleaner on it](https://github.com/osresearch/heads-wiki/blob/master/Clean-the-ME-firmware.md#how-to-disabledeactive-most-of-it) and reflash on it.
  - x230: On first boot of x230-flash Heads, `mount-usb && flash.sh /media/coreboot.rom` (12mb rom generated with make BOARD=x230)
- On first boot of Heads, the user is given an error that `Heads couldn't find any GPG keys in the keyring`.  He is invited to `Add a GPG key to the running BIOS` from a USB drive if the GPG card was preowned externally, or to `Factory reset, own GPG card, generate and add key into running BIOS` to do the whole process from inside or Heads, with generated key and trustdb being backuped onto USB drive under `gpg_keys` directory for further use. 
new doc 2019-01-19 15 13 25_1
  - The GPG generated public key is valid for one year and that validity can be extended by the user. the key `/gpg_keys/public.key` can be published online and is a 4096bits RSA key, ready to use, combined with it's private key, secured inside of the GPG card. By selecting `Factory reset, own GPG card, generate and add key into running BIOS`, the user also injected otrust.txt inside of the rom, which is imported at each Heads reboot, ultimately trusting generated user key.
- On next reboot, the user/OEM is warned that `Heads couldn't generate the TOTP code` and is invited to `Reset the TPM` if he didn't own it yet with is own passphrase, or to `Generate new TOTP/HOTP secret`, to seal BIOS integrity attestation into the TPM and into GPG card previously chosen admin password.
new doc 2019-01-19 15 13 25_2
- _At this point, the BIOS attestation with HOTP/TOTP is done and any modification done to it will detected._
new doc 2019-01-19 15 13 25_3
- OEM/User then installs QubesOS from Heads (Whiptail) from `Advanced Settings-> Other Boot Options -> USB boot` which validates integrity of the ISO with companion `.asc` checksum file, while Heads validates that checksum authenticity with [Heads distro key included signature files](https://github.com/osresearch/heads/tree/master/initrd/etc/distro/keys)
new doc 2019-01-19 15 13 25_4
new doc 2019-01-19 15 13 25_5
new doc 2019-01-19 15 13 25_6
- On next reboot when the OEM/user will attempt to select a default boot option from normal boot menu, he will receive an error saying that `The file containing hashes for /boot is missing` and be prompted to update the checksums and sign them with it's GPG card and his user chosen password.
new doc 2019-01-18 15 06 28_19
new doc 2019-01-18 15 06 28_20
- On next reboot, the user/OEM will be prompted to select a default boot option and Make it the default. _**By doing so, the user/OEM will enter Disk Recovery key passphrase chosen at OS installation, and will be prompted to choose a distinct Disk Unlock key, that will be released in LUKS slot 1 (instead of slot 0 by the Recovery Key) by the TPM only if it attests BIOS integrity. This is important for data privacy. By doing so, whenever the user would  be filmed typing is Unlock disk passphrase (instead of typing otherwise his Disk Recovery Key), an attacker would only be able to access the disk content on this computer and won't be able to clone that disk content and use that same passphrase to unlock disk content. Consequently, if the BIOS is tampered with/upgraded/modified, selecting the default boot option and typing Disk Unlock Key passphrase won't work, which is an additional safeguard against external tampering.**_
new doc 2019-01-18 15 06 28_22
new doc 2019-01-18 15 06 28_23
new doc 2019-01-18 15 06 28_24
new doc 2019-01-18 15 06 28_25
new doc 2019-01-18 15 06 28_26
new doc 2019-01-18 15 06 28_27
new doc 2019-01-18 15 06 28_28

**OEM specifics**:
- After having done all the previous steps, the OEM enters recovery shell and enters: `mount -o remount,rw /boot && touch /boot/oem && umount /boot && reboot`
- The OEM then ships the GPG card and hardware separately.

**OEM-> user devices transfer of ownership:**
- The user plugs in the LibremKey/NitroKey Pro v2 and boots the hardware and gets the following:
new doc 2019-01-18 15 06 28_1
new doc 2019-01-18 15 06 28_2
new doc 2019-01-18 15 06 28_3
new doc 2019-01-18 15 06 28_4
new doc 2019-01-18 15 06 28_5
new doc 2019-01-18 15 06 28_6
new doc 2019-01-18 15 06 28_7
new doc 2019-01-18 15 06 28_8
new doc 2019-01-18 15 06 28_9
new doc 2019-01-18 15 06 28_10
new doc 2019-01-18 15 06 28_11
new doc 2019-01-18 15 06 28_12
new doc 2019-01-18 15 06 28_13
new doc 2019-01-18 15 06 28_14
new doc 2019-01-18 15 06 28_15
new doc 2019-01-18 15 06 28_16
new doc 2019-01-18 15 06 28_17
new doc 2019-01-18 15 06 28_18
new doc 2019-01-18 15 06 28_19
new doc 2019-01-18 15 06 28_20
new doc 2019-01-18 15 06 28_21
new doc 2019-01-18 15 06 28_22
new doc 2019-01-18 15 06 28_23
new doc 2019-01-18 15 06 28_24
new doc 2019-01-18 15 06 28_25
new doc 2019-01-18 15 06 28_26
new doc 2019-01-18 15 06 28_27
new doc 2019-01-18 15 06 28_28

@marmarek @mfc @andrewdavidwong @kylerankin @osresearch @flammit : Your input would be more then welcome. I'm going to use this PoC as it is for a start, but would love it to be mainstreamed and [merge efforts to ease remote support of users](https://github.com/SkypLabs/my-qubes-os-formula/issues/6).

This seals my PoC to go forward with the [QubesOS Hardware Certification](https://github.com/QubesOS/qubes-issues/issues/1771#issuecomment-449543084).
I await from you  guys what kind of partnership/collaboration can be constructed with you guys to facilitate QubesOS adoption to a larger user base, mostly those who really need it.

**Comment by tlaurion** [comment]
replaced by #511

**Pull request closed without merging**